src/raft.rs

```diff
@@ -884,16 +994,132 @@ impl<T: Storage> Raft<T> {
         self.r.send_append_aggressively(to, pr, &mut self.msgs)
     }
 
+    // Find an appropriate agent.
+    // If found an appropriate one, mark the corresponding message's type as
+    // MsgGroupBroadcast and return true. If not found, return false.
+    fn select_agent_for_bcast_group(&self, msgs: &mut [Message]) -> bool {
+        for msg in msgs {
+            let peer_id = msg.to;
+            let is_voter = self.prs().conf().voters().contains(peer_id);
+            // Agent must be voter and recently active.
+            if !is_voter || !self.is_recent_active(peer_id) {
+                continue;
+            }
+            msg.set_msg_type(MessageType::MsgGroupBroadcast);
+            return true;
+        }
+        false
+    }
+
+    fn merge_msg_group(&mut self, mut group: Vec<Message>) {
+        // Do not need to merge if group size is less than two.
+        if group.len() < 2 {
+            self.msgs.append(&mut group);
+            return;
+        }
+        // Send messages directly if no appropriate agent in this broadcast group.
+        if !self.select_agent_for_bcast_group(&mut group) {
+            self.msgs.append(&mut group);
+            return;
+        }
+
+        // Record forward information
+        let mut forwards: Vec<Forward> = Vec::default();
+        let mut mark = 0;
+        for (idx, msg) in group.iter().enumerate() {
+            // MsgAppend sent to other peers in this group only reserve basic
+            // forward information.
+            if msg.get_msg_type() != MessageType::MsgGroupBroadcast {
+                let forward = Forward {
+                    to: msg.to,
+                    log_term: msg.log_term,
+                    index: msg.index,
+                    ..Default::default()
+                };
+                forwards.push(forward);
+            } else {
+                mark = idx;
+            }
+        }
+        // Attach forward information to MsgGroupbroadcast and send it.
+        group[mark].set_forwards(forwards.into());
+        for msg in group {
+            if msg.get_msg_type() == MessageType::MsgGroupBroadcast {
+                self.msgs.push(msg);
+                return;
+            }
+        }
+    }
+
     /// Sends RPC, with entries to all peers that are not up-to-date
     /// according to the progress recorded in r.prs().
     pub fn bcast_append(&mut self) {
         let self_id = self.id;
+        let leader_group_id = self
```

Comment:
Can this be queried only when follower replication is enabled?

Reply:
There are two conditions that we should use leader replication. One is that follower replication is disabled. The other is that follower replication is enabled but leader's group id is unknown. So I want to reuse the code of iterating progress and sending append.

Reply:
```
let mut use_leader_replication = self.is_leader();
let leader_group_id = 0;
if !use_leader_replication {
    leader_group_id = query;
    use_leader_replciation = leader_group_id == 0;
}
if use_leader_replication {
    return;
}   
```